packages/ember-data/lib/system/snapshot.js

```diff
@@ -0,0 +1,298 @@
+  */
+  id: function() {
+    return this._id;
+  },
+
+  /**
+    Returns the current value of a belongsTo relationship.
+
+    Example
+
+    ```javascript
+    var post = store.createRecord('post', { title: 'Hello World' });
+    var comment = store.createRecord('comment', { body: 'Lorem ipsum', post: post });
+    var snapshot = comment._createSnapshot();
+
+    snapshot.belongsTo('post'); // => DS.Snapshot of post
+    ```
+
+    Calling `belongsTo` will return a new Snapshot as long as there's any
+    data available, such as an ID. If there's no data available `belongsTo` will
+    return undefined.
+
+    Note: Relationships are loaded lazily and cached upon first access.
+
+    @method belongsTo
+    @param {string} keyName
+    @return {DS.Snapshot|undefined} A snapshot of a belongsTo relationship or undefined
+  */
+  belongsTo: function(keyName) {
+    var relationship, snapshot, inverseRecord;
+
+    if (!(keyName in this._belongsToRelationships)) {
+      relationship = this.record._relationships[keyName];
+
+      if (!(relationship && relationship.relationshipMeta.kind === 'belongsTo')) {
+        throw new Ember.Error("Model '" + Ember.inspect(this.record) + "' has no belongsTo relationship named '" + keyName + "' defined.");
+      }
+
+      inverseRecord = get(relationship, 'inverseRecord');
+      if (inverseRecord) {
+        snapshot = inverseRecord._createSnapshot();
+      }
+
+      this._belongsToRelationships[keyName] = snapshot;
+    }
+
+    return this._belongsToRelationships[keyName];
+  },
+
+  /**
+    Returns the current value of a hasMany relationship.
+
+    `hasMany` takes an optional hash of options as a second parameter,
+    currently supported options are:
+
+   - `ids`: set to `true` if you only want the IDs of the related items to be
+      returned.
+
+    Example
+
+    ```javascript
+    var post = store.createRecord('post', { title: 'Hello World', comments: [2, 3] });
+    var snapshot = post._createSnapshot();
+
+    snapshot.hasMany('comments'); // => [DS.Snapshot, DS.Snapshot]
+    snapshot.hasMany('comments', { ids: true }); // => [2, 3]
+    ```
+
+    Note: Relationships are loaded lazily and cached upon first access.
+
+    @method hasMany
+    @param {string} keyName
+    @param {Object} [options]
+    @return {Array} An array of snapshots of a hasMany relationship
+  */
+  hasMany: function(keyName, options) {
+    var ids = options && options.ids;
+    var snapshots = [];
+    var relationship, members, hasMany;
+
+    if (!(keyName in this._hasManyRelationships)) {
+      relationship = this.record._relationships[keyName];
+
+      if (!(relationship && relationship.relationshipMeta.kind === 'hasMany')) {
+        throw new Ember.Error("Model '" + Ember.inspect(this.record) + "' has no hasMany relationship named '" + keyName + "' defined.");
+      }
+
+      members = get(relationship, 'members');
+      members.forEach(function(member) {
```

Comment:
My reasoning for this is to cache the hasMany on first access, no matter if you want just the IDs or the snapshots.

Imagine this:

```javascript
var snapshot = post.snapshot();
var ids = snapshot.hasMany('comments', { ids: true });
var comments = snapshot.hasMany('comments');
```

If we don't create the snapshots and cache them on the first access we could have a scenario when `ids` and `comments` in the example above doesn't match.

Thoughts?